microcosm_flask/namespaces.py

```diff
@@ -0,0 +1,154 @@
+"""
+Defines namespaces for operations that follow various API conventions.
+
+In conjunction with the `Operation` enum, defines a naming convention for HTTP endpoints,
+which in turn provides a discovery mechanism API routes.
+
+"""
+from urllib import urlencode
+from urlparse import urljoin
+
+from flask import request, url_for
+
+from microcosm_flask.naming import (
+    collection_path_for,
+    instance_path_for,
+    name_for,
+    relation_path_for,
+    singleton_path_for,
+)
+from microcosm_flask.operations import Operation
+
+
+class Namespace(object):
+    """
+    Encapsulates the namespace for one or more operations.
+
+    Each fully qualified operation can be viewed as a subject, verb, and optional object.
+
+    The `Operation` enum defines the legal verbs (according to various conventions); this
+    object encapsulates the rest.
+
+    """
+
+    def __init__(self, subject, object_=None, path="", version=None):
+        self.subject = subject
+        self.object_ = object_
+        self.path = path
+        self.version = version
+
+    @property
+    def object_ns(self):
+        """
+        Create a new namespace for the current namespace's object value.
+
+        """
+        return Namespace(
+            path=self.path,
+            subject=self.object_,
+            object_=None,
+            version=self.version,
+        )
+
+    @property
+    def subject_name(self):
+        return name_for(self.subject)
+
+    @property
+    def object_name(self):
+        return name_for(self.object_)
+
+    @property
+    def collection_path(self):
+        return self.path + collection_path_for(self.subject)
+
+    @property
+    def instance_path(self):
+        return self.path + instance_path_for(self.subject)
+
+    @property
+    def relation_path(self):
+        return self.path + relation_path_for(self.subject, self.object_)
+
+    @property
+    def singleton_path(self):
+        return self.path + singleton_path_for(self.subject)
+
+    def endpoint_for(self, operation):
```

Comment:
One of the motivations of the `Namespace` abstraction is that it gives us a good place to inject alternate logic for endpoint naming.

Example: we've talked a bit about having low-level and high-level APIs (for the same resource). If we had such a thing for a delete operation, we'd currently end up with **two** endpoint names called `foo.delete`, which Flask will not allow. The path forward is to give the `Namespace` another member variable and generate different endpoint names based on some prefix (e.g. `x.foo.delete` and `y.foo.delete`).